### PR TITLE
infra: setup staging deploy

### PR DESCRIPTION
**Pull request by carissa-tang**
## Notion ticket link

[Setup Staging Environment](https://www.notion.so/uwblueprintexecs/Setup-Staging-Environment-1ba835eb693c4bdd90616a789d092a61?pvs=4)

## Implementation description
* Set up staging and prod sites on Firebase
* Updated GitHub Actions workflows with targets to support
   1. Staging preview deployment: for all FE PR changes, a temporary "preview" staging link will be generated
   2. Staging deployment: on merge of FE changes to staging, changes are deployed to jump-math-staging.web.app
   3. Prod deployment: on merge of FE changes to main, changes are deployed to jumpmath.web.app
* Deployed backend to Heroku (https://jump-math-98edf.herokuapp.com/)
   * Note: GraphQL playground is turned off
   * Will be deployed on merge of BE changes to staging and main
* Updated CORS_ALLOW_LIST with new URLs
* Set missing secrets in GitHub Action Secrets
* Updated Firebase Action URL from localhost to https://jump-math-staging.web.app (for email actions)

## Steps to test
1. Play around on https://jump-math-staging--pr128-carissa-setup-stagin-e2ql18pf.web.app/ ! You can test the backend by logging in, making accounts, removing users, etc.
* Note there is a flaw in the email verification flow, since it redirects to https://jump-math-staging.web.app which is not live yet. You can replace the link with the preview link (i.e. https://jump-math-staging--pr128-carissa-setup-stagin-e2ql18pf.web.app/...firebase stuff...) to make it work for now.

## What should reviewers focus on?
* TODO: set up a new MongoDB cluster for staging / prod
   * I have set staging / prod to use the same cluster as our current one for local development, but we should separate them later on
* Using GitHub's student credits rn (https://www.heroku.com/github-students, $13 USD per month x 12 months) - should last a year (??) but will need to discuss long term solution
   * Currently have my personal credit card as the temporary payment method on Heroku

## Checklist
- [ ] My PR name is descriptive and in imperative tense
- [ ] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [ ] I have run the appropriate linter(s)
- [ ] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: .github/workflows/deploy-preview.yml [hunk 1]
```diff
@@ -1,29 +1,29 @@
...
 
 jobs:
   build_and_preview:
-    if: "${{ github.event.pull_request.head.repo.full_name == github.repository && github.base_ref == 'main' }}"
+    if: '${{ github.event.pull_request.head.repo.full_name == github.repository }}'
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v2
       - run: echo "REACT_APP_BACKEND_URL=${{ secrets.PREVIEW_BACKEND_URL }}" > .env
-      - run: echo "REACT_APP_OAUTH_CLIENT_ID=${{ secrets.DEV_OAUTH_CLIENT_ID }}" >> .env
```
Comment (carissa-tang) [inline .github/workflows/deploy-preview.yml:22]: used for google login - unneeded

File: frontend/.firebaserc [hunk 2]
```diff
@@ -1,5 +1,18 @@
...
   "projects": {
-    "default": "uw-blueprint-starter-code"
-  }
+    "default": "jump-math-98edf"
+  },
+  "targets": {
+    "jump-math-98edf": {
+      "hosting": {
+        "prod": [
+          "jumpmath"
```
Comment (carissa-tang) [inline frontend/.firebaserc:9]: corresponds to our two hosting sites on firebase

image

File: frontend/package.json [hunk 3]
```diff
@@ -40,7 +40,7 @@
     "react-json-schema": "^1.2.2",
     "react-jsonschema-form": "^1.8.1",
     "react-router-dom": "^5.2.0",
-    "react-scripts": "4.0.2",
+    "react-scripts": "^5.0.1",
```
Comment (carissa-tang) [inline frontend/package.json]: update to latest react-scripts

File: frontend/package.json [hunk 4]
```diff
@@ -47,8 +47,8 @@
     "web-vitals": "^1.0.1"
   },
   "scripts": {
-    "start": "react-scripts start",
-    "build": "react-scripts build",
+    "start": "react-scripts --openssl-legacy-provider start",
```
Comment (carissa-tang) [inline frontend/package.json:51]: to address build error for upgraded node version in staging https://stackoverflow.com/questions/69665222/node-js-17-0-1-gatsby-error-digital-envelope-routinesunsupported-err-os

File: frontend/src/components/assessment-creation/AssessmentsTable.tsx [hunk 5]
```diff
@@ -29,4 +29,5 @@ const AssessmentsTable = ({
 
   return <Table headers={headers} rows={rows} />;
 };
+
```
Comment (carissa-tang) [inline frontend/src/components/assessment-creation/AssessmentsTable.tsx:32]:
```suggestion
```

File: backend/Procfile [hunk 6]
```diff
@@ -1 +1 @@
-web: node migrate up && node build/server.js
```
Comment (carissa-tang) [inline backend/Procfile:1]: i believe this is for postgres db only - gives error for us

File: frontend/Dockerfile [hunk 7]
```diff
@@ -1,4 +1,4 @@
-FROM node:14.19-slim
+FROM node:18.14.1-slim
```
Comment (carissa-tang) [inline frontend/Dockerfile:1]: enforce new node version locally to ensure `--openssl-legacy-provider` flag is valid

**Review by joyce-shi — APPROVED** [review]
LGTM!

**Pull request merged**